src/uu/tail/src/tail.rs

```diff
@@ -356,49 +636,543 @@ pub fn uu_app<'a>() -> Command<'a> {
         )
 }
 
-/// Continually check for new data in the given readers, writing any to stdout.
-fn follow<T: BufRead>(readers: &mut [(T, &String)], settings: &Settings) -> UResult<()> {
-    if readers.is_empty() || !settings.follow {
-        return Ok(());
+fn follow(files: &mut FileHandling, settings: &mut Settings) -> UResult<()> {
+    let mut process = platform::ProcessChecker::new(settings.pid);
+
+    let (tx, rx) = channel();
+
+    // Watcher is implemented per platform using the best implementation available on that
+    // platform. In addition to such event driven implementations, a polling implementation
+    // is also provided that should work on any platform.
+    // Linux / Android: inotify
+    // macOS: FSEvents / kqueue
+    // Windows: ReadDirectoryChangesWatcher
+    // FreeBSD / NetBSD / OpenBSD / DragonflyBSD: kqueue
+    // Fallback: polling every n seconds
+
+    // NOTE:
+    // We force the use of kqueue with: features=["macos_kqueue"].
+    // On macOS only `kqueue` is suitable for our use case because `FSEvents`
+    // waits for file close util it delivers a modify event. See:
+    // https://github.com/notify-rs/notify/issues/240
+
+    let mut watcher: Box<dyn Watcher>;
+    if settings.use_polling || RecommendedWatcher::kind() == WatcherKind::PollWatcher {
+        settings.use_polling = true; // We have to use polling because there's no supported backend
+        let config = notify::poll::PollWatcherConfig {
+            poll_interval: settings.sleep_sec,
+            ..Default::default()
+        };
+        watcher = Box::new(notify::PollWatcher::with_config(tx, config).unwrap());
+    } else {
+        let tx_clone = tx.clone();
+        match notify::RecommendedWatcher::new(tx) {
+            Ok(w) => watcher = Box::new(w),
+            Err(e) if e.to_string().starts_with("Too many open files") => {
+                // NOTE: This ErrorKind is `Uncategorized`, but it is not recommended to match an error against `Uncategorized`
+                // NOTE: Could be tested with decreasing `max_user_instances`, e.g.:
+                // `sudo sysctl fs.inotify.max_user_instances=64`
+                show_error!(
+                    "{} cannot be used, reverting to polling: Too many open files",
+                    text::BACKEND
+                );
+                settings.exit_code = 1;
+                settings.use_polling = true;
+                let config = notify::poll::PollWatcherConfig {
+                    poll_interval: settings.sleep_sec,
+                    ..Default::default()
+                };
+                watcher = Box::new(notify::PollWatcher::with_config(tx_clone, config).unwrap());
+            }
+            Err(e) => panic!("called `Result::unwrap()` on an `Err` value: {:?}", &e),
```

Comment:
It feels weird to manually construct the error message of unwrap if there is no actual `unwrap` call. Could you make it something more descriptive instead? (I.e. "notify::RecommendedWatcher::new returned an unexpected error" or something like that)